src/Data/Array/Accelerate/AST/Idx.hs

```diff
@@ -76,12 +76,12 @@ newtype Idx env t = UnsafeIdxConstructor { unsafeRunIdx :: Int }
 
 {-# COMPLETE ZeroIdx, SuccIdx #-}
 
-pattern ZeroIdx :: forall env envt t. () => (envt ~ (env, t)) => Idx envt t
-pattern ZeroIdx <- (\x -> (idxToInt x, unsafeCoerce @(envt :~: envt) @(envt :~: (env, t)) Refl) -> (0, Refl))
+pattern ZeroIdx :: forall envt t. () => forall env. (envt ~ (env, t)) => Idx envt t
+pattern ZeroIdx <- (\x -> (idxToInt x, unsafeCoerce @_ @(envt :~: (_, t)) Refl) -> (0, Refl))
```

Comment:
Ah, I didn't realise `env` would be in scope there!

Reply:
Aha, so I guess `env` is not in scope at all! We just suggestively name a new type variable `env`, and GHC later infers that the two `env` types must be equal (and that one does not escape the scope of the other).

Reply:
I suppose so, yes. Actually that makes me think that my version is _less_ clear, because it suggests a type relationship that doesn't actually hold.

I'm now actually almost considering naming that thing `kaas` just to illustrate that it doesn't matter. (well perhaps something like `doesn'tMatter`)

Reply:
Something like the underscore perhaps, that I had before? ;)
I think `env` is fine, since the relationship does hold (just not due to `scopedTypeVariables`, but simple unification).

Reply:
True lol. Yes now in hindsight I think your original version was actually fine. Or at least, my version is not better :p